crates/re_log_types/src/data_table_batcher.rs

```diff
@@ -0,0 +1,787 @@
+            new.flush_num_rows = s
+                .parse()
+                .map_err(|err| DataTableBatcherError::ParseConfig {
+                    name: Self::ENV_FLUSH_NUM_ROWS,
+                    value: s.clone(),
+                    err: Box::new(err),
+                })?;
+        }
+
+        Ok(new)
+    }
+}
+
+#[test]
+fn data_table_batcher_config() {
+    // Detect breaking changes in our environment variables.
+    std::env::set_var("RERUN_FLUSH_TICK_SECS", "0.3");
+    std::env::set_var("RERUN_FLUSH_NUM_BYTES", "42");
+    std::env::set_var("RERUN_FLUSH_NUM_ROWS", "666");
+
+    let config = DataTableBatcherConfig::from_env().unwrap();
+
+    let expected = DataTableBatcherConfig {
+        flush_tick: Duration::from_millis(300),
+        flush_num_bytes: 42,
+        flush_num_rows: 666,
+        ..Default::default()
+    };
+
+    assert_eq!(expected, config);
+}
+
+// ---
+
+/// Implements an asynchronous batcher that coalesces [`DataRow`]s into [`DataTable`]s based upon
+/// the thresholds defined in the associated [`DataTableBatcherConfig`].
+///
+/// ## Multithreading and ordering
+///
+/// [`DataTableBatcher`] can be cheaply clone and used freely across any number of threads.
+///
+/// Internally, all operations are linearized into a pipeline:
+/// - All operations sent by a given thread will take effect in the same exact order as that
+///   thread originally sent them in, from its point of view.
+/// - There isn't any well defined global order across multiple threads.
+///
+/// This means that e.g. flushing the pipeline ([`Self::flush_blocking`]) guarantees that all
+/// previous data sent by the calling thread has been batched; no more, no less.
+///
+/// ## Shutdown
+///
+/// The batcher can only be shutdown dropping all instances of it, at which point it will
+/// automatically take care of flushing any pending data that might remain in the pipeline.
+///
+/// Shutting down cannot ever block, unless the output channel is bounded and happens to be full
+/// (see [`DataTableBatcherConfig::max_tables_in_flight`]).
+#[derive(Clone)]
+pub struct DataTableBatcher {
+    inner: Arc<DataTableBatcherInner>,
+}
+
+// NOTE: The receiving end of the command stream as well as the sending end of the table stream are
+// owned solely by the batching thread.
+struct DataTableBatcherInner {
+    /// The one and only entrypoint into the pipeline: this is _never_ cloned nor publicly exposed,
+    /// therefore the `Drop` implementation is guaranteed that no more data can come in while it's
+    /// running.
+    tx_cmds: Sender<Command>,
+    rx_tables: Receiver<DataTable>,
+    cmds_to_tables_handle: Option<std::thread::JoinHandle<()>>,
+}
+
+impl Drop for DataTableBatcherInner {
+    fn drop(&mut self) {
+        // NOTE: The command channel is private, if we're here, nothing is currently capable of
+        // sending data down the pipeline.
+        self.tx_cmds.send(Command::Shutdown).ok();
```

Comment:
I think we want to `drop self.rx_tables` *before* we call this send to avoid a deadlock?

Basically, if we are using a bounded channels and rx_tables isn't being drained by any listeners, then `tx_table.send(table)` could be blocking the batching thread, preventing it from ever processing the Shutdown.  Dropping rx_tables first should at least call tx_table to either error *or* it means some other outstanding sender thread is still holding a receiver and then at least it isn't our problem.

Reply:
Yeah the docs mention this:
> /// Shutting down cannot ever block, unless the output channel is bounded and happens to be full
> /// (see [`DataTableBatcherConfig::max_tables_in_flight`]).

My thought process being that if the user deliberately configures the channel sizes to be bounded, then they should expect that the system can and will block at any (inconvenient) time if they don't consume as needed.

Eh, I guess we can be extra polite...